File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> prefixFreq = new HashMap<>();
        int sum = 0;
        int result = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            if (sum == k) {
                result++;
            }
            if (prefixFreq.containsKey(sum - k)) {
                result += prefixFreq.get(sum - k);
            }
            prefixFreq.put(sum, prefixFreq.getOrDefault(sum, 0) + 1);
        }
        return result;
    }
}


// 1, 1, 1, k = 2
// output: 2
// 1, 2, 3

// 3,  4,  7, -2,  2,  1,  4,  2   k = 7
// 3,  7, 14, 12, 14, 15, 19, 21

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefixSum = defaultdict(int)
        prefixSum[0] = 1
        runningSum = 0
        result = 0
        for num in nums:
            runningSum += num
            target = runningSum - k
            if target in prefixSum.keys():
                result += prefixSum[target]
            if runningSum in prefixSum.keys():
                prefixSum[runningSum] += 1
            else:
                prefixSum[runningSum] = 1
        return result